File: exercise2.hs
module Main where

-- Run with: cat exercise2_input.txt | ./exercise 2

data SubmarinePosition = SubmarinePosition Int Int Int deriving (Show)

foldInstructionToSubmarine :: SubmarinePosition -> (String, Int) -> SubmarinePosition
foldInstructionToSubmarine (SubmarinePosition hozPos depth aim) ("forward", amount) = SubmarinePosition (hozPos + amount) depth aim
foldInstructionToSubmarine (SubmarinePosition hozPos depth aim)("down", amount) = SubmarinePosition hozPos (depth + amount) aim
foldInstructionToSubmarine (SubmarinePosition hozPos depth aim) ("up", amount) = SubmarinePosition hozPos (depth - amount) aim

foldInstructionToSubmarinePart2 :: SubmarinePosition -> (String, Int) -> SubmarinePosition
foldInstructionToSubmarinePart2 (SubmarinePosition hozPos depth aim) ("forward", amount)  = SubmarinePosition (hozPos + amount) (depth + (amount * aim)) aim
foldInstructionToSubmarinePart2 (SubmarinePosition hozPos depth aim) ("down", amount) = SubmarinePosition hozPos depth (aim + amount)
foldInstructionToSubmarinePart2 (SubmarinePosition hozPos depth aim) ("up", amount) = SubmarinePosition hozPos depth (aim - amount)

multiplied (SubmarinePosition x y _) = x * y

linesToProperTuple :: [String] -> [(String, Int)]
linesToProperTuple [] = []
linesToProperTuple (x:y:xs) = [(x, read y :: Int)] ++ linesToProperTuple xs

main = do
  content <- getContents
  let contentLines = words content
  let instructions = linesToProperTuple contentLines
  let submarine = foldl foldInstructionToSubmarine (SubmarinePosition 0 0 0) instructions
  print submarine
  putStrLn $ "Multiplied part one = " ++ show (multiplied submarine)
  let submarinePartTwo = foldl foldInstructionToSubmarinePart2 (SubmarinePosition 0 0 0) instructions
  print submarinePartTwo
  putStrLn $ "Multiplied part two = " ++ show (multiplied submarinePartTwo)

File: exercise2.py
def part_one(actions: [(str, int)]):
    horizontal_pos = 0
    depth = 0
    for (action, amount) in actions:
        if action == 'forward':
            horizontal_pos = horizontal_pos + amount
        elif action == 'down':
            depth = depth + amount
        elif action == 'up':
            depth = depth - amount
    print(f"""Part one. Horizontal pos: {horizontal_pos},
              depth: {depth}, multiplied {horizontal_pos * depth}""")


def part_two(actions: [(str, int)]):
    horizontal_pos = 0
    depth = 0
    aim = 0
    for (action, amount) in actions:
        if action == 'forward':
            horizontal_pos = horizontal_pos + amount
            depth = depth + (aim * amount)
        elif action == 'down':
            aim = aim + amount
        elif action == 'up':
            aim = aim - amount
    print(f"""Part two. Horizontal pos: {horizontal_pos}, depth: {depth},
              aim: {aim}, multiplied {horizontal_pos * depth}""")


def main():
    with open('exercise2_input.txt', 'r') as input_file:
        # First split in newines, then split each line on
        # space and convert amount to int
        actions = [(x.split(' ')[0], int(x.split(' ')[1]))
                   for x in input_file.read().split("\n") if x != '']
    part_one(actions)
    part_two(actions)


if __name__ == '__main__':
    main()
